changes_test.go

```diff
@@ -656,11 +980,49 @@ func (s *changesSuite) TestFromData(c *gc.C) {
 		err = data.Verify(nil)
 		c.Assert(err, jc.ErrorIsNil)
 
-		// Check that the changes are what we expect.
+		// Retrieve the changes, and convert them to a sequence of records.
 		changes := bundlechanges.FromData(data)
-		b, err := json.MarshalIndent(changes, "", "  ")
+		records := make([]record, len(changes))
+		for i, change := range changes {
+			r := record{
+				Id:       change.Id(),
+				Requires: change.Requires(),
+				Method:   change.Method(),
+				GUIArgs:  change.GUIArgs(),
+			}
+			switch change := change.(type) {
+			case *bundlechanges.AddCharmChange:
+				r.Args = change.Args
+			case *bundlechanges.AddMachineChange:
+				r.Args = change.Args
+			case *bundlechanges.AddRelationChange:
+				r.Args = change.Args
+			case *bundlechanges.AddServiceChange:
+				r.Args = change.Args
+			case *bundlechanges.AddUnitChange:
+				r.Args = change.Args
+			case *bundlechanges.SetAnnotationsChange:
+				r.Args = change.Args
+			default:
+				c.Fatalf("unsupported change type %T", change)
+			}
+			records[i] = r
+		}
+
+		// Output the records for debugging.
+		b, err := json.MarshalIndent(records, "", "  ")
 		c.Assert(err, jc.ErrorIsNil)
-		c.Logf("obtained changes: %s", b)
-		c.Assert(changes, jc.DeepEquals, test.expected)
+		c.Logf("obtained records: %s", b)
+
+		// Check that the obtained records are what we expect.
+		c.Assert(records, jc.DeepEquals, test.expected)
 	}
 }
+
+type record struct {
```

Comment:
Perhaps define this at the start of the file rather than the end?

Also a doc comment might be good for such a generically named type.
For example:

    // record holds expected information about the contents
    // of a change value.